sherpa/models/tests/test_model.py

```diff
@@ -1079,6 +1102,101 @@ def calc(self, p, *args, **kwargs):
         return numpy.ones(args[0].size)
 
 
+def get_cache_classes():
+    """This is a function because we want to conditionally
+    include an XSPEC model. Within a function, we can simply
+    pass that if XSPEC is not available.
+    """
+    cls_list = [DoNotUseModel, Polynom1D]
+
+    try:
+        from sherpa.astro.xspec import XSphabs
+        cls_list.append(XSphabs)
+    except ImportError:
+        pass
+
+    return cls_list
+
+
+@pytest.mark.parametrize('cls', get_cache_classes())
+def test_cache_uses_instance_attributes(cls):
+    """Check that the cache uses the instance attributes.
+
+    This tests both the real code (e.g. ArithmetricModels)
+    but also the DoNotUseModel class which we defined above
+    just for testing, because it previously failed just in that
+    test class in a hard-to-debug way.
+    """
+    mdl = cls("some-name")
+
+    # cache is an int, not a mutable object, so it's OK if it's the same
+    # as the class attribute, unless we set it - then it ought to be different.
+    mdl.cache = 2234
+
+    for attr in ["cache", "_cache", "_cache_ctr"]:
+        assert hasattr(mdl, attr)
+        if hasattr(cls, attr):
+            assert id(getattr(mdl, attr)) != id(getattr(cls, attr))
+
+
+def test_cache_reset_when_size_changes():
+    """Check the cache is reset when the cache size changes."""
+
+    mdl = Polynom1D()
+    mdl.cache = 2
+
+    x = numpy.arange(2, 10, 1.5)
+    mdl(x)
+
+    assert len(mdl._cache) == 1
+    assert mdl._cache_ctr['check'] == 1
+
+    mdl.cache = 3
+    assert len(mdl._cache) == 0
+    assert mdl._cache_ctr['check'] == 0
+
+    mdl(x)
+    assert len(mdl._cache) == 1
+    assert mdl._cache_ctr['check'] == 1
+
+
+def test_caching_not_used_when_set_to_zero():
+    """Check the cache is not used when the cache size is zero."""
+
+    mdl = Polynom1D()
+    mdl.cache = 0
+
+    x = numpy.arange(2, 10, 1.5)
+    mdl(x)
+
+    assert len(mdl._cache) == 0
+    # We checked, but found nothing (because the cache is empty)
+    assert mdl._cache_ctr['check'] == 1
+
+
+def test_cache_not_used_in_fit():
+    """Check the cache is not used in a fit when `fit=False`.
+
+    How do we do that without relying too much on the internal
+    implementation?
+    """
+    mdl = Const1D('con1')
+    mdl.c0 = 1
+    dat = Data1D('data', numpy.arange(4), 2 * numpy.ones(4), numpy.ones(4))
+    fit = Fit(dat, mdl)
+    res = fit.fit()
+    assert mdl.c0.val == pytest.approx(2)
+    assert len(mdl._cache) == 5
+
+    mdl.cache_clear()
+    mdl.c0 = 1
+    res = fit.fit(cache=False)
+    assert mdl.c0.val == pytest.approx(2)
+    assert len(mdl._cache) == 0
+    # There are no values in the cache while in the case with cache=True (the default)
```

Comment:
In the comment, would it be clearer to indicate 'while caching is enabled' rather than 'cache=True'?